### PR TITLE
Use Map for rule search and add benchmark

### PR DESCRIPTION
**Pull request by lupomontero**
This PR includes the changes proposed by @gugu's in #302 as well as on other minor tweaks (like only iterating over the rules once instead of twice on module load).

This increases performance (ops/sec) by a factor of over 100!!

The main change is using a `Map` object and indexing it by `punySuffix`. This allows searching by `punySuffix` without having to iterate over the thousands of rules.

```
npm run benchmark

> psl@1.13.0 benchmark
> node --experimental-vm-modules --no-warnings benchmark/suite.js

┌────────────────────────────────────┬─────────┬──────┬──────────────┐
│ (index)                            │ ops/sec │ ± %  │ runs sampled │
├────────────────────────────────────┼─────────┼──────┼──────────────┤
│ [source] psl#isValid               │ 367385  │ 3.76 │ 85           │
│ [source] psl#parse                 │ 396952  │ 1.22 │ 93           │
│ [source] psl#parse invalid domain  │ 429242  │ 1.25 │ 92           │
│ [v1.13.0] psl#isValid              │ 3306    │ 1.62 │ 87           │
│ [v1.13.0] psl#parse                │ 3402    │ 2.07 │ 89           │
│ [v1.13.0] psl#parse invalid domain │ 3230    │ 1.53 │ 87           │
└────────────────────────────────────┴─────────┴──────┴──────────────┘
┌──────────────────────────┬────────┬─────────┬──────────────┐
│ (index)                  │ source │ v1.13.0 │ diff         │
├──────────────────────────┼────────┼─────────┼──────────────┤
│ psl#isValid              │ 367385 │ 3306    │ '111.13x up' │
│ psl#parse                │ 396952 │ 3402    │ '116.68x up' │
│ psl#parse invalid domain │ 429242 │ 3230    │ '132.89x up' │
└──────────────────────────┴────────┴─────────┴──────────────┘
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2,32 +2,34 @@ import Punycode from 'punycode/punycode.js';
 
 import rules from './data/rules.js';
 
-
 var internals = {};
```
Comment (mfdebian) [inline index.js:7]: Is there a specific reason for this whitespace here? It's such a tiny change but I'd just leave it like this
```suggestion

```

File: index.js [hunk 2]
```diff
@@ -2,32 +2,34 @@ import Punycode from 'punycode/punycode.js';
...
-// Check if given string ends with `suffix`.
-//
-internals.endsWith = function (str, suffix) {
-
-  return str.indexOf(suffix, str.length - suffix.length) !== -1;
-};
+    return map;
+  },
+  new Map(),
+);
```
Comment (mfdebian) [inline index.js:32]: Brilliant.

I tried to find a case/scenario were it could "break", I tried to check with incredibly long URLs like:

```
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb.cccccccccccccccccccccccccccccccccccccccccccccccccccccc.dddddddddddddddddddddddddddddddddddddddddddddddddddddd.com
```

But even then I still got `'64.70x up'` on the benchmark run.

I also tried `very.complex.nested.subdomain.example.co.uk` and it also works like a charm.

I feel like I've got nothing of substance to add, this just works great, you've reduced the complexity significantly. :clap: :clap:

File: index.js [hunk 3]
```diff
@@ -36,25 +38,15 @@ internals.endsWith = function (str, suffix) {
...
-    //if (memo) {
-    //  var memoSuffix = Punycode.toASCII(memo.suffix);
-    //  if (memoSuffix.length >= punySuffix.length) {
-    //    return memo;
-    //  }
-    //}
-    return rule;
-  }, null);
+  }
+  return null;
```
Comment (mfdebian) [inline index.js:49]: I also tried to think about this one, I've read about how sometimes inverted lookup is supposed to be faster but I don't really know, LGTM!

File: benchmark/suite.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    fn: () => {
+      module.parse('google.com');
+    }
+  });
+
+  suite.add('psl#parse invalid domain', {
+    fn: () => {
+      module.parse('google.comp');
+    }
+  });
```
Comment (mfdebian) [inline benchmark/suite.js:24]: I believe we should add more examples to the suite just so we're testing edge cases:

1.- Extremely long URLs
2.- Complex URLs
3.- The ones with non-conventional characters
4.- Wildcard domains?
5.- Domains with hyphens, underscores
6.- Extremely nested subdomains

Let me know what you think, I can do that and send a PR

File: benchmark/suite.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  const mod = new vm.SourceTextModule(text);
+
+  await mod.link(() => { });
+  await mod.evaluate();
+
+  return mod.namespace;
+};
+
+const main = async () => {
+  const compareToVersion = 'v1.13.0';
```
Comment (mfdebian) [inline benchmark/suite.js:113]: I tried my best not to let this bother me. :rofl: 

I understand it allows us to test to different versions, in this case this would be the same as just having `url` be `https://unpkg.com/psl@latest/dist/psl.mjs` but we might want to test other cases, I get it.

I guess I should first ask how are we thinking about treating this file, is it something we're going to be updating on every release? are we always going to be testing against the same value we could find as `version` on our `package.json`? Or maybe I'm reading too much into it :shrug: idk, I'm trying to think of a way to keep this without having a human remembering to change this value.

**Review by mfdebian — APPROVED** [review]
Looks Amazing To Me :star: 

I guess I don't really have much to add, sadly, but I ran it, tried some edge cases and it works like a charm :sparkles: 

Literally a HUGE improvement in terms of complexity :clap:

File: index.js (hunk 1 above)
Comment (lupomontero) [reply index.js:7]: No specific reason... it just slipped through...

File: index.js (hunk 2 above)
Comment (lupomontero) [reply index.js:32]: Having you test it locally and check the code gives me confidence to merge this :muscle:

File: benchmark/suite.js (hunk 5 above)
Comment (lupomontero) [reply benchmark/suite.js:113]: So far this script has only been used for manually benchmarking against the precending version (the current stable release) in the context of this PR (which is specifically about performance, and the original PR #302 was responsible for adding the script). In the future we could automate the benchmarking to always automatically compare new changes to the _current_ published version, but would only look into it when and if we feel it would add value (when addressing performance issues)..

Comment (lupomontero) [reply benchmark/suite.js:113]: It could be a good idea to pass the version we want to compare to as a command line argument. That way, by default it would compare the local source code vs the _latest_ published version.

```sh
npm run benchmark
```

Alternatively, you would also be able to compare to a specific published version (ie: `v1.13.0`):

```sh
npm run benchmark -- v1.13.0
```

To achieve this I think we'd only need to change the following line:

https://github.com/lupomontero/psl/blob/440248a70c5c42aaa87a9b9cca349f506d367482/benchmark/suite.js#L113

Into this:

```js
const compareToVersion = process.argv[2] || 'latest';
```

Comment (lupomontero) [reply benchmark/suite.js:113]: I've created a separate issue for this so we can proceed with this PR before adding more changes. See #335

**Comment by lupomontero** [comment]
Thanks for the review @mfdebian :yellow_heart:
